### PR TITLE
Add diag_embed to ATen and torch

### LINKED ISSUE #12160: [feature request] Batched (n-1)-D to n-D matrix_diag
should be easy to implement with `diagonal`.

### PR DESCRIPTION
**Pull request by t-vi**
Fixes: #12160

### CODE REVIEW

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -2358,6 +2363,67 @@ def parse_kwargs(desc):
...
+        (main diagonal).
+    dim1 (int, optional): first dimension with respect to which to
+        take diagonal. Default: -2.
+    dim2 (int, optional): second dimension with respect to which to
+        take diagonal. Default: -1.
+
+Example::
+
+    >>> x = torch.randn(2, 3)
+    tensor([[[ 1.5410,  0.0000,  0.0000],
```
Comment (SsnL) [inline torch/_torch_docs.py]: hmm this example tensor doesn't look like that it is from randn.

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -2358,6 +2363,67 @@ def parse_kwargs(desc):
     tensor([ 1.2252,  0.5002,  0.6248,  2.0139])
 """)
 
+add_docstr(torch.matrix_diag,
+           r"""
+matrix_diag(input, offset=0, dim1=-2, dim2=-1) -> Tensor
+
+Creates a diagonal matrix from :attr:`input` by using the last
```
Comment (SsnL) [inline torch/_torch_docs.py]: This first sentence is a bit unclear. It usually creates more than 1 matrix, or even the created tensors are not matrices. I also have some questions about the name and API raised above.

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -1026,6 +1026,9 @@
 
 - func: matmul_out(Tensor result, Tensor self, Tensor other) -> Tensor
 
+- func: matrix_diag(Tensor self, int64_t offset=0, int64_t dim1=-2, int64_t dim2=-1) -> Tensor
+  variants: function, method
```
Comment (SsnL) [inline aten/src/ATen/native/native_functions.yaml]: Should it be a method as well? (I'm not sure)

**Review by ssnl — COMMENTED** [review]
I'm still a bit unsure about the name and the API. 

On name: it works on higher rank tensors than 2. So `matrix` is a bit misleading.
On API: Wouldn't it be nice if users can specify which dimension is used as the diagonal?

File: torch/_torch_docs.py (hunk 1 above)
Comment (t-vi) [reply torch/_torch_docs.py]: Oops. Something went wrong copypasting this.

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (t-vi) [reply aten/src/ATen/native/native_functions.yaml]: I think moving from diagonal and back is quite common, so some people will prefer it to have a method. But if you think it's safer to just add the function until someone complains, I'll go with that.

**Comment by t-vi** [comment]
On the naming: Indeed, I agree that matrix_diag is a not-so-nice name. I would have called that `diagonal_tensor` if I had to come up with my own name. As such I'll be more than happy to change the name.

On the API: I think that diagonal and this function should match, this is why I took the last dimension. I would expect this to cover the by far most common case.
When I implemented diagonal, I thought it was strange to always put the diagonal at the end, but that is what numpy did. I haven't, however, heard anyone complain, so adding another parameter might be overthinking the applicability of this function. One could always add a "source dimension" keyword argument later (unless that does bad things to the C++ API). Again, I'll gladly add an argument (and would suggest to do the same in torch.diagonal).

**Comment by t-vi** [comment]
> Is there really no term for making a diagonal matrix out of a vector?

I'm sure the category theory-inclined will love diagonal_morphism or diagonal_embedding, but we would have to keep  people searching for these to look at Wikipedia.

**Comment by apaszke** [comment]
Yeah, "diagonal embedding" seems to be a good term (the morphism could really go both ways, so it remains unclear). It's only a bit verbose. How about `diag_embed`/`embed_diag`?

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (SsnL) [reply aten/src/ATen/native/native_functions.yaml]: Sure let's keep it as this.

**Comment by SsnL** [comment]
One vote from me for `embed_diag`, although keeping the same `diag_` prefix would be nice.

**Comment by apaszke** [comment]
Yeah, `diag_*` is definitely good for discoverability, but it's a much worse method name 😕 Can't really say which one is better, but I think `diag_embed` is better. We can't expect people to go through 400 functions we provide.

**Comment by fmassa** [comment]
About the naming: I suppose numpy doesn't have an equivalent function? FWIW `matrix_diag` is the name if the function in TF

**Comment by apaszke** [comment]
`matrix_diag` is really not a good name... The function works with any tensors, and the name doesn't make it clear if it takes the diagonal, or fills the diagonal of a higher-dim tensor.

**Comment by t-vi** [comment]
So I renamed to diag_embed.

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -1106,6 +1106,69 @@ def parse_kwargs(desc):
...
+The argument :attr:`offset` controls which diagonal to consider:
+
+- If :attr:`offset` = 0, it is the main diagonal.
+- If :attr:`offset` > 0, it is above the main diagonal.
+- If :attr:`offset` < 0, it is below the main diagonal.
+
+The size of the new matrix will be calculated to make the specified diagonal
+of the size of the last input dimension.
+Note that for :attr:`offset` other than :math:`0`, the order of :attr:`dim1`
+and :attr:`dim2` matters, exchaning them is equivalent to changing the
```
Comment (SsnL) [inline torch/_torch_docs.py]: nit: exchanging

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -1106,6 +1106,69 @@ def parse_kwargs(desc):
     tensor([ 0.0255, 0.1374])
 """)
 
+add_docstr(torch.diag_embed,
+           r"""
+diag_embed(input, offset=0, dim1=-2, dim2=-1) -> Tensor
+
+Creates a tensor with entries on a diagonal from :attr:`input` by
```
Comment (SsnL) [inline torch/_torch_docs.py]: This first sentence confuses me... Maybe `` Creates a tensor whose diagonals of certain 2D planes (specified by :attr:`dim1` and :attr:`dim2`) are filled by :attr:`input`. ``?

**Review by ssnl — COMMENTED** [review]
Code looks clean, as always. However, I do have some comments on the doc.

**Comment by t-vi** [comment]
I think both CI failures are unrelated to the patch.

**Comment by t-vi** [comment]
@SsnL @apaszke anything I can do to move this forward?

**Review by soumith — APPROVED** [review]

**Review by ssnl — COMMENTED** [review]
I am so sorry. I had comments typed but forgot to hit "submit review" button....

File: torch/_torch_docs.py [hunk 6]
```diff
@@ -1106,6 +1106,69 @@ def parse_kwargs(desc):
     tensor([ 0.0255, 0.1374])
 """)
 
+add_docstr(torch.diag_embed,
+           r"""
+diag_embed(input, offset=0, dim1=-2, dim2=-1) -> Tensor
+
+Creates a tensor whose diagonals of certain 2D planes (specified by
+:attr:`dim1` and :attr:`dim2`) are filled by :attr:`input`. By default, the
```
[the remark this answers is not in the document]
Comment (t-vi) [reply torch/_torch_docs.py]: Thanks! I put "by default" at the end, I hope that is OK.

**Review by ssnl — APPROVED** [review]
Thank you so much!

**Comment by SsnL** [comment]
hmm could you rebase? It seems that the base commit is in bad state. Thanks!

**Comment by vadimkantorov** [comment]
If this is in, how about re-considering [`torch.eye_like`](https://github.com/pytorch/pytorch/issues/9244) with two new arguments dim1 and dim2 (which will make it clear, across which dimensions the diagonal is filled with 1's)

**Pull request closed without merging**